src/index/Vocabulary.h

```diff
@@ -85,6 +85,104 @@ class PrefixComparator {
   const Vocabulary<S>* _vocab;
 };
 
+/**
+ * \brief Comparator for std::strings that optionally supports
+ * case-insensitivity.
+ *
+ * If the constructor is called with ignoreCase=false it is an ordinary string
+ * compare If ignoreCase=True the operator behaves as follows:
+ *
+ * - compare the strings according to their lowercase version
+ * - in case the lowercase versions are equal, return the order of the original
+ * strings
+ * - (including case)
+ *
+ * This gives us a strict ordering on strings.
+ */
+class StringSortComparator {
+ public:
+  StringSortComparator(bool ignoreCase = false) : _ignoreCase(ignoreCase) {}
+
+  bool getIgnoreCase() const { return _ignoreCase; }
+
+  bool operator()(const std::string& a, const std::string& b) const {
+    if (!_ignoreCase) {
+      return a < b;
+    } else {
+      // TODO<Johannes>
+      // BUG<Johannes>
+      // We have to make sure that the literals are all in contiguous space to
+      // make this work.
+      // TODO<Johannes> Ideally we want to have this also when doing
+      // case-insensitive compare, but it currently breaks the prefix
+      // compression (there we really need ordering by correct bytes)
+      auto splitA = extractComparable(a);
+      auto splitB = extractComparable(b);
+      if (splitA.isLiteral != splitB.isLiteral) {
+        // only one is a literal, compare by the first character to separate
+        // datatypes.
+        return a < b;
+      }
+      return caseInsensitiveCompare(splitA, splitB);
+    }
+  }
+
+  struct SplitVal {
+    SplitVal(bool lit, std::string_view v, std::string_view l)
+        : isLiteral(lit), val(v), langtag(l) {}
+    bool isLiteral;
+    std::string_view val;
+    std::string_view langtag;
+  };
+
+  static SplitVal extractComparable(const std::string& a) {
+    std::string_view res = a;
+    bool isLiteral = false;
+    std::string_view langtag;
+    if (ad_utility::startsWith(res, "\"")) {
+      // note: this only works for valid SPARQL literals with at least two
+      // quotation marks but this is ensured for all the vocabulary words.
+      isLiteral = true;
+      auto pos = res.rfind('\"');
+      // this should also be fine if there is no langtag (pos == size() ok
+      // according to cppreference.com
+      langtag = res.substr(pos + 1);
+      res.remove_suffix(res.size() - pos);
+      res.remove_prefix(1);
+    }
+    return {isLiteral, res, langtag};
+  }
+
+  static bool caseInsensitiveCompare(const SplitVal& a, const SplitVal& b) {
+    const auto result =
+        std::mismatch(a.val.cbegin(), a.val.cend(), b.val.cbegin(),
+                      b.val.cend(), [](const auto& lhs, const auto& rhs) {
+                        return tolower(lhs) == tolower(rhs);
```

Comment:
This too should use an Unicode aware `tolower()`

Reply:
I think we currently have to map the whole string to lowercase using utf-8 and then do the mismatch thing because in c++ the whole "variable length" business is not well-supported. (as long as the performance is ok, I will argue about this below).